### PR TITLE
Update camera2 app

### PR DESCRIPTION
**Pull request by jiawen**

### CODE REVIEW

**Comment by steven-johnson** [comment]
LGTM

**Comment by jiawen** [comment]
Please take another look. I wrote a new YuvBufferT class instead of "yet another buffer_t wrapper" and a bunch of utility functions.

File: apps/HelloAndroidCamera2/jni/BufferTFunctions.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        }
+    } else {
+        // Copy element by element.
+        int32_t srcElementStrideBytes = src.stride[0] * src.elem_size;
+        int32_t dstElementStrideBytes = dst.stride[0] * dst.elem_size;
+        int32_t srcRowStrideBytes = src.stride[1] * src.elem_size;
+        int32_t dstRowStrideBytes = dst.stride[1] * dst.elem_size;
+
+        // Just do array assignment if elements are small enough.
+        // It's slightly ridiculous that we have all these versions, but oh well.
```
Comment (steven-johnson) [inline apps/HelloAndroidCamera2/jni/BufferTFunctions.cpp]: Could collapse into a local templated function if you want something a little terser.

File: apps/HelloAndroidCamera2/jni/HalideFilters.cpp [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+#define LOGE(...) __android_log_print(ANDROID_LOG_ERROR, "native", __VA_ARGS__)
+
+#include "AndroidBufferUtilities.h"
+#include "BufferTFunctions.h"
+#include "YuvFunctions.h"
+#include "deinterleave.h"
+#include "edge_detect.h"
+#include "HalideRuntime.h"
+
+#define DEBUG 1
```
Comment (steven-johnson) [inline apps/HelloAndroidCamera2/jni/HalideFilters.cpp]: Do you mean to check this in with DEBUG defined?

File: apps/HelloAndroidCamera2/jni/HalideFilters.cpp [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+// Extern functions from the Halide runtime that are not exposed in
+// HalideRuntime.h.
+extern "C" int halide_host_cpu_count();
```
Comment (steven-johnson) [inline apps/HelloAndroidCamera2/jni/HalideFilters.cpp]: If they aren't defined they might change... not saying we can't do this but it's not a good precedent for something included with Halide :-)

File: apps/HelloAndroidCamera2/jni/HalideFilters.cpp [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+    buffer_t srcLuma = src->luma();
+    buffer_t dstLuma = dst->luma();
+    int64_t t1 = halide_current_time_ns();
+    int err = edge_detect(&srcLuma, &dstLuma);
+    if (err != halide_error_code_success) {
+        LOGE("edge_detect failed with error code: %d", err);
+    }
+
+    int64_t t2 = halide_current_time_ns();
```
Comment (steven-johnson) [inline apps/HelloAndroidCamera2/jni/HalideFilters.cpp]: You probably want this logging section to be ifdef DEBUG or similar.

File: apps/HelloAndroidCamera2/jni/HalideFilters.cpp [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            src->luma().extent[0], src->luma().extent[1],
+            src->chromaU().extent[0], src->chromaU().extent[1],
+            src->chromaV().extent[0], src->chromaV().extent[1],
+            dst->luma().extent[0], dst->luma().extent[1],
+            dst->chromaU().extent[0], dst->chromaU().extent[1],
+            dst->chromaV().extent[0], dst->chromaV().extent[1]);
+        return false;
+    }
+
+    static bool first_call = true;
```
Comment (steven-johnson) [inline apps/HelloAndroidCamera2/jni/HalideFilters.cpp]: You probably want this logging section to be ifdef DEBUG or similar.

File: apps/HelloAndroidCamera2/jni/YuvBufferT.cpp [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    chromaV_.stride[1] = chromaVRowStrideBytes;
+    chromaV_.min[0] = 0;
+    chromaV_.min[1] = 0;
+    chromaV_.elem_size = 1;
+
+    // See if chroma is stored according to a well known format.
+    chromaStorage_ = ChromaStorage::kOther;
+    // U and V must have the same extents and strides.
+    if (equalExtents(chromaU_, chromaV_) && equalStrides(chromaU_, chromaV_)) {
+        // If strides are exactly 2, check of they are interleaved.
```
Comment (steven-johnson) [inline apps/HelloAndroidCamera2/jni/YuvBufferT.cpp]: s/of/if/

File: apps/HelloAndroidCamera2/jni/YuvFunctions.h [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+#include "HalideRuntime.h"
+
+#include "YuvBufferT.h"
+
+// Returns true if a and b have the same extent.
+bool equalExtents(const YuvBufferT &a, const YuvBufferT &b);
+
+// Returns true if src and dst have the same extent and the copy was
+// successful, false otherwise.
+bool copy2D(const YuvBufferT &src, const YuvBufferT &dst);
```
Comment (steven-johnson) [inline apps/HelloAndroidCamera2/jni/YuvFunctions.h]: Note that result == false can leave dst in an inconsistent state (some copied, some not)

File: apps/HelloAndroidCamera2/jni/BufferTFunctions.cpp (hunk 1 above)
Comment (jiawen) [inline apps/HelloAndroidCamera2/jni/BufferTFunctions.cpp]: Thought about it and decided it wasn't worth it. It'd be a function with way too many parameters. Punt?

File: apps/HelloAndroidCamera2/jni/HalideFilters.cpp (hunk 2 above)
Comment (jiawen) [inline apps/HelloAndroidCamera2/jni/HalideFilters.cpp]: Yes, this didn't change from before.

File: apps/HelloAndroidCamera2/jni/HalideFilters.cpp (hunk 3 above)
Comment (jiawen) [inline apps/HelloAndroidCamera2/jni/HalideFilters.cpp]: I just copy/pasted the boilerplate from Andrew's camera1 example. What's the right thing to do here?

File: apps/HelloAndroidCamera2/jni/YuvBufferT.cpp (hunk 6 above)
Comment (jiawen) [inline apps/HelloAndroidCamera2/jni/YuvBufferT.cpp]: Done.

File: apps/HelloAndroidCamera2/jni/YuvFunctions.h (hunk 7 above)
Comment (jiawen) [inline apps/HelloAndroidCamera2/jni/YuvFunctions.h]: Done.

**Comment by jiawen** [comment]
What's the usual workflow? I'll push additional commits to the branch and squash before the final merge?

File: apps/HelloAndroidCamera2/jni/HalideFilters.cpp (hunk 4 above)
Comment (jiawen) [inline apps/HelloAndroidCamera2/jni/HalideFilters.cpp]: Eh, it's pretty standard Android with LOGD and LOGE.

File: apps/HelloAndroidCamera2/jni/HalideFilters.cpp (hunk 5 above)
Comment (jiawen) [inline apps/HelloAndroidCamera2/jni/HalideFilters.cpp]: Eh, it's pretty standard Android with LOGD and LOGE.

**Comment by abadams** [comment]
Push additional commits. Squashing is not required.

File: apps/HelloAndroidCamera2/jni/HalideFilters.cpp (hunk 3 above)
Comment (abadams) [inline apps/HelloAndroidCamera2/jni/HalideFilters.cpp]: Not every OS uses host_cpu_count, so it's not always available, so it's not exposed in HalideRuntime.h. Android does have it though. We could make it part of the API and add a windows implementation (getenv("NUMBER_OF_PROCESSORS")).

halide_current_time_ns could probably go in HalideRuntime.h too

Comment (abadams) [inline apps/HelloAndroidCamera2/jni/HalideFilters.cpp]: I think these are OK to extern declare for now.

**Comment by steven-johnson** [comment]
LGTM

**Comment by abadams** [comment]
Taking it for a spin...

**Comment by abadams** [comment]
Smooth as silk. Thanks Kevin.

**Pull request merged**